docs/alerts.md

```diff
@@ -0,0 +1,183 @@
+# Alert Handling
+
+## Overview
+
+### Requirements
+
+The following are considered requirements for the alert package to both support and/or implement.
+
+* It must be able to consume data from an arbitrary query and evaluate a level function on each row in each table.
+* It must be possible to store this result in a system bucket called a `status bucket`.
+* It must be able to evaluate a `notification rule` based on the output of the previous or from the data stored in the status bucket.
+* The notification must send messages to third party endpoints when the rule indicates.
+* The notification events must be capable of being logged into a `notification bucket`.
+
+### Query Lifecycle
+
+The query lifecycle is tangentially related to alerts.
+At the moment, the flux engine is currently only used for running batch requests.
+This means that there is only one table for each group key that goes through the pipeline.
+
+The alert package as below does not assume this.
+It is designed so the functions will work regardless of whether run in batch mode or if the query were being run with each new row triggering a new table.
+For ease of reading, the example tables will be shown as if only one table were shown.
+
+### Considerations
+
+While designing the alert package, we have looked at the previous work done in this field that was included in [Kapacitor](https://docs.influxdata.com/kapacitor/v1.5/nodes/alert_node/).
+
+It is important to support the same workflows that Kapacitor allows.
+
+## Alerts and Notifications
+
+Alerts and notifications is comprised of two parts: checking the input and notifying a third party about the output of the check.
+
+For this reason, the Flux library will treat these as separate concepts.
+When evaluating an input stream to determine the alert level, we will need to evaluate each row to determine if that row matches into one of 4 possible alert levels: `ok`, `info`, `warn`, and `crit`.
+
+Each of these levels will be represented by an integer to make it easier to compare the levels.
+There will also be an additional level to represent that a check was not able to determine the status or the status was not yet known: `unknown`.
+
+They will be mapped as the following:
+
+|level|value|
+|-----|-----|
+|ok|0|
+|info|1|
+|warn|2|
+|crit|3|
+|unknown|-1|
+
+### Check
+
+The `check` function will take an input and evaluate each row to determine the level.
+The function will have a predicate for each specified level and will output a level for the highest matching check.
+One output table will be produced for each input table.
+The output tables will have the same schema, but with one additional column that is `_level:int`.
```

Comment:
Check will need to output a `_check` column that contains the name/id of the check. 

Additionally users need to be able to add arbitrary tags to the check output. This can easily be done with a map function, but we should make it easy to do. Perhaps with a `tags` parameter to the `check` function.

Reply:
I'll add a `name` parameter which will add a value to the group key for the name of the check.

What do you mean by arbitrary tags? To clarify, the current design would keep the existing tags from the input. So it might be better to have a `set()` function that allows setting a value in a column or setting a value and adding it to the group key.

I'm not sure this needs to be part of the behavior of the `check()` function though because check will pass through the existing data and the existing group key.

Reply:
@jsternberg Yes using the `set` function is what I had in mind. Basically users will be able to inject new tags on their data as part of the check. This allows them to categorize their checks. This is similar to Kapacitor's topic system but less rigid.